File: hw/codena/u3/hw21Interfaces/Python/AnimalFarmSystem/model/farm_animal.py
from abc import ABC, abstractmethod
from datetime import date

class FarmAnimal(ABC):
    def __init__(self, id, breed, born_on_date, weight, slaughter_house, product, cuts):
        self.id = id
        self.breed = breed
        self.born_on_date = born_on_date
        self.weight = weight
        self.slaughter_house = slaughter_house
        self.product = product
        self.cuts = cuts

    def get_age_in_months(self):
        if self.born_on_date is None:
            raise ValueError("born_on_date cannot be None")

        current_date = date.today()

        months = (
            (current_date.year - self.born_on_date.year) * 12
            + (current_date.month - self.born_on_date.month)
        )

        if current_date.day < self.born_on_date.day:
            months -= 1

        return months
    
    @abstractmethod
    def feed(self, food):
        pass

File: hw/codena/u3/hw21Interfaces/JavaScript/AnimalFarmSystem/model/FarmAnimal.js
class FarmAnimal {
    constructor(id, breed, bornOnDate, weight, slaughterHouse, product, cuts) {

        if (new.target === FarmAnimal) {
            throw new Error("FarmAnimal is an abstract class.");
        }

        this.id = id
        this.breed = breed
        this.bornOnDate = bornOnDate
        this.weight = weight
        this.slaughterHouse = slaughterHouse
        this.product = product
        this.cuts = cuts
    }


    getAgeInMonths() {
        if (this.bornOnDate == null) {
            throw new Error("bornOnDate cannot be null");
        }

        const currentDate = new Date();

        let months =
            (currentDate.getFullYear() - this.bornOnDate.getFullYear()) * 12 +
            (currentDate.getMonth() - this.bornOnDate.getMonth());

        if (currentDate.getDate() < this.bornOnDate.getDate()) {
            months--;
        }

        return months;
    }


    feed(food) {
        throw new Error("The method feed() must be implemented.");
    }


}